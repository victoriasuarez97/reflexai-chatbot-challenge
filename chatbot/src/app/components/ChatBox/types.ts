type Props = {
    user: string
}

export type ChatBoxType = (props: Props) => JSX.Element
