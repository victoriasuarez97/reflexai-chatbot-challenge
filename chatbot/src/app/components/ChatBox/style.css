#chatContainer {
    height: 70vh;
    width: 50vw;
    overflow-y: auto;
    display: flex;
    flex-direction: column;
    animation: fadeIn 0.2s;
}

form {
    margin-top: auto;
    height: 40px;
}
