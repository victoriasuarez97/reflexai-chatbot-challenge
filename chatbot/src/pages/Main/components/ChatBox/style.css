input#userMessage {
    padding: 8px 16px;
    border-radius: 8px;
    border-color: grey;
    outline-color: pink;
    width: 100%;
    height: 45px;
}