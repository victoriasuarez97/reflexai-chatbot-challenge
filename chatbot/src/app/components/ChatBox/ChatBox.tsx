'use client'

import { Avatar, Box, Flex, IconButton, Text, TextField, Toast } from "gestalt"
import { useChat } from 'ai/react';
import { ChatBoxType } from "./types";

import 'gestalt/dist/gestalt.css';
import './style.css'

export const ChatBox: ChatBoxType = ({ user }) => {
    const { messages, input, handleInputChange, handleSubmit, error } = useChat({
        initialMessages: [
            {
                id: '1',
                role: 'assistant',
                content: `Hello ${user} =^.^= I'm your catssistant!`
            }
        ],
    });

    if (error) return (
        <Flex
            justifyContent="center"
            alignItems="center"
            width="100%"
            height="100%"
        >
            <Toast
                type="error"
                text="Something went wrong. Please try again later."
                dismissButton={{ onDismiss: () => window.location.reload() }}
            />
        </Flex>
    )

    return (
        <Box borderStyle="sm" rounding={3} padding={6} margin={6} id='chatContainer'>
            <Flex direction="column">
                {
                    messages.map(({ id, role, content }) => (
                        <Box key={id} marginTop={4} marginBottom={4}>
                            <Flex alignItems="center" wrap={false} width='100%'>
                                <Box>
                                    {role === 'user'
                                        ? <Avatar name={user} size='md' />
                                        : <Avatar src='/robot.png' name='Catbot' verified size='md' />
                                    }
                                </Box>
                                <Box marginStart={3}>
                                    <Text weight='bold'>{role === 'user' ? user : 'Catbot'}</Text>
                                    <Text>{content}</Text>
                                </Box>
                            </Flex>
                        </Box>
                    ))
                }
            </Flex>
            <form onSubmit={handleSubmit}>
                <Flex alignItems="center" wrap={false} width='100%'>
                    <label htmlFor="userMessage" />
                    <input
                        id='userMessage'
                        onChange={handleInputChange}
                        value={input}
                        placeholder="Ask me anything"
                        autoComplete="off"
                    />
                    <IconButton icon="send" accessibilityLabel="Send message" type="submit" />
                </Flex>
            </form>
        </Box>
    )
}