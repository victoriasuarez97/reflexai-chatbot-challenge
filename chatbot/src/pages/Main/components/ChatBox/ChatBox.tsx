'use client'

import { Avatar, Box, Flex, IconButton, Text, Toast } from "gestalt"
import { useChat } from 'ai/react';

import 'gestalt/dist/gestalt.css';
import './style.css'

export const ChatBox = () => {
    const { messages, input, handleInputChange, handleSubmit, error } = useChat();

    if (error) return (
        <Flex
            justifyContent="center"
            alignItems="center"
            width="100%"
            height="100%"
        >
            <Toast
                type="error"
                text="Something went wrong. Please try again later."
                dismissButton={{ onDismiss: () => {}}}
            />
        </Flex>
    )

    return (
        <Box borderStyle="sm" rounding={3} padding={6} margin={6} id='chatContainer'>
            <Flex direction="column">
                {
                    messages.map(({ id, role, content }) => (
                        <Box key={id} marginTop={4} marginBottom={4}>
                            <Flex alignItems="center" wrap={false} width='100%'>
                                <Box>
                                    {role === 'user'
                                        ? <Avatar name='User' size='md' />
                                        : <Avatar src='/robot.png' name='Catbot' verified size='md' />
                                    }
                                </Box>
                                <Box marginStart={3}>
                                    <Text weight='bold'>{role === 'user' ? 'User' : 'Catbot'}</Text>
                                    <Text>{content}</Text>
                                </Box>
                            </Flex>
                        </Box>
                    ))
                }
            </Flex>
            <form onSubmit={handleSubmit}>
                <Flex alignItems="center" wrap={false} width='100%'>
                    <label htmlFor="userMessage" />
                    <input
                        id='userMessage'
                        onChange={handleInputChange}
                        value={input}
                        placeholder="Ask me anything"
                        autoComplete="off"
                    />
                    <IconButton icon="send" accessibilityLabel="Send message" type="submit" />
                </Flex>
                <Text size='100'>Please don&apos;t forget to be polite always</Text>
            </form>
        </Box>
    )
}