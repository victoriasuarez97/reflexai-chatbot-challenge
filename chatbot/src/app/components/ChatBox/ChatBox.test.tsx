import { fireEvent, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ChatBox } from './ChatBox'
import { useChat, Message } from 'ai/react'

const spyHook = { useChat }

const useChatSpy = jest.spyOn(spyHook, 'useChat')
const mockHandleSubmit = jest.fn()
const mockHandleInputChange = jest.fn()
const baseResponse = {
    messages: [{ role: 'system' as Message['role'], content: "Hello Username =^.^= I'm your catssistant!", id: '90873489273984' }],
    handleSubmit: mockHandleSubmit,
    error: undefined,
    append: jest.fn(),
    reload: jest.fn(),
    stop: jest.fn(),
    setMessages: jest.fn(),
    input: '',
    setInput: jest.fn(),
    handleInputChange: mockHandleInputChange,
    isLoading: false
}

afterEach(() => {
  jest.clearAllMocks()
})

describe('ChatBox test', () => {
    it('should initial message from bot', () => {
        const { getByText } = render(<ChatBox user='Username' />)

        expect(getByText("Hello Username =^.^= I'm your catssistant!")).toBeInTheDocument()
    })

    it('should find placeholder when there is no message typed in input', () => {
        const { getByPlaceholderText } = render(<ChatBox user='Username' />)

        expect(getByPlaceholderText('Ask me anything')).toBeInTheDocument()
    })

    it('should be able to type and submit text', async () => {
        useChatSpy.mockReturnValue(baseResponse)

        const { getByPlaceholderText } = render(<ChatBox user='Username'/>)

        const input = getByPlaceholderText('Ask me anything')

        userEvent.type(input, 'hello caatbot')

        fireEvent.submit(input)
    })

    it('should match snapshot when there is a start and completion', () => {
        useChatSpy.mockReturnValue({
            ...baseResponse,
            messages: [
                { role: 'user' as Message['role'], content: 'hello caatbot', id: '90873489273984' },
                { role: 'system' as Message['role'], content: 'hello there!', id: '90873489273985' }
            ]
        })

        const { container } = render(<ChatBox user='Username'/>)

        expect(container).toMatchSnapshot()
    })

    it('should be able to find error text when service fails', async () => {
        useChatSpy.mockReturnValue({
            ...baseResponse,
            error: { name: 'New Error', message: 'Something failed!!' }
        })

        const { getByText, getByPlaceholderText, rerender } = render(<ChatBox user='Username'/>)

        const input = getByPlaceholderText('Ask me anything')

        userEvent.type(input, 'hey')

        fireEvent.submit(input)

        rerender(<ChatBox user='Username'/>)

        await waitFor(() => 
            expect(getByText('Something went wrong. Please try again later.')).toBeInTheDocument())
    })
})