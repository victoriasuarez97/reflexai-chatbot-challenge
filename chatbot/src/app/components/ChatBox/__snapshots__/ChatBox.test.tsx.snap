// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBox test should match snapshot when there is a start and completion 1`] = `
<div>
  <div
    class="zI7 iyn Hsu"
    id="container"
  >
    <div
      class="LJB PKX Pw5 QDA Tte daS dxm fBv fte s2n zI7 iyn Hsu"
      id="chatContainer"
    >
      <div
        class="jzS un8 C9i TB_"
      >
        <div
          class="KO4 RDc zI7 iyn Hsu"
        >
          <div
            class="KS5 hs0 un8 C9i TB_"
            style="width: 100%;"
          >
            <div
              class="zI7 iyn Hsu"
            >
              <div
                class="INd XiG zI7 iyn Hsu"
                data-test-id="gestalt-avatar-svg"
                style="width: 48px; height: 48px;"
              >
                <div
                  class="Pj7 sLG XiG INd m1e"
                >
                  <div
                    class="XiG zI7 iyn Hsu"
                    style="padding-bottom: 100%;"
                  >
                    <img
                      alt="Catbot"
                      class="hCL kVc L4E MIw"
                      fetchpriority="auto"
                      loading="auto"
                      src="/robot.png"
                    />
                  </div>
                  <div
                    class="KPc MIw ojN Rym p6V QLY"
                  />
                </div>
                <div
                  class="MIw zI7 iyn Hsu"
                  style="width: 25%; height: 25%; min-width: 12px; min-height: 12px; bottom: 4%; right: 4%;"
                >
                  <div
                    class="INd imm zI7 iyn Hsu"
                    style="width: 100%; height: 100%;"
                  >
                    <svg
                      aria-hidden="true"
                      aria-label=""
                      class="g_1 gUZ U9O kVc"
                      height="100%"
                      role="img"
                      viewBox="0 0 24 24"
                      width="100%"
                    >
                      <path
                        d="M12 0a12 12 0 1 0 0 24 12 12 0 0 0 0-24M6.37 11.61a1.25 1.25 0 0 1 1.76 0l2.37 2.36 5.37-5.35a1.25 1.25 0 0 1 1.76 1.76L10.5 17.5l-4.13-4.13a1.25 1.25 0 0 1 0-1.76"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Pyg zI7 iyn Hsu"
            >
              <div
                class="tBJ dyH iFc sAJ X8m zDA IZT H2s"
              >
                Catbot
              </div>
              <div
                class="tBJ dyH iFc sAJ X8m zDA IZT swG"
              >
                Hello Username =^.^= I'm your catssistant!
              </div>
            </div>
          </div>
        </div>
      </div>
      <form>
        <div
          class="KS5 hs0 un8 C9i TB_"
          style="width: 100%;"
        >
          <label
            for="userMessage"
          />
          <input
            autocomplete="off"
            id="userMessage"
            placeholder="Ask me anything"
            value=""
          />
          <button
            aria-label="Send message"
            class="HEm adn yQo lnZ wsz"
            tabindex="0"
            type="submit"
          >
            <div
              class="rYa kVc adn yQo S9z qrs BG7"
            >
              <div
                class="x8f INd _O1 KS5 mQ8 OGJ"
                style="height: 48px; width: 48px;"
              >
                <svg
                  aria-hidden="true"
                  aria-label=""
                  class="Hn_ Uvi gUZ U9O kVc"
                  height="20"
                  role="img"
                  viewBox="0 0 24 24"
                  width="20"
                >
                  <path
                    d="m.46 2.43-.03.03c-.4.42-.58 1.06-.28 1.68L3 10.5 16 12 3 13.5.15 19.86c-.3.62-.13 1.26.27 1.67l.05.05c.4.38 1 .56 1.62.3l20.99-8.5q.28-.12.47-.3l.04-.04c.68-.71.51-2-.51-2.42L2.09 2.12Q1.79 2 1.49 2q-.61.01-1.03.43"
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
